pkg/bridgeclient/websocket_message_handler.go

```diff
@@ -0,0 +1,153 @@
+	return "web-socket"
+}
+func (t *WebSocketMessageHandler) StartMessageHandler(clientID string) error {
+	urlSplit := strings.SplitAfterN(t.serverURL, "://", 2)
+	urlObject := url.URL{
+		Scheme: "ws",
+		Host:   urlSplit[1],
+		Path:   fmt.Sprintf("/bridge-server/1/message-queue/%s/ws", clientID),
+	}
+	websocketURL := urlObject.String()
+	log.WithField("websocketURL", websocketURL).Info("Using websocket transport")
+
+	conn, _, err := websocket.DefaultDialer.Dial(websocketURL, nil)
+	if err != nil {
+		log.WithError(err).WithField("url", websocketURL).Error("Failed to connect to websocket")
+		return err
+	}
+	go t.subscribeAndSendMessageToCloud(conn,clientID)
+	go t.ReadWSFromCloud(conn)
+	return nil
+}
+func (t *WebSocketMessageHandler) StopMessageHandler() {
+	//TODO Should probably do something here
+}
+func (t *WebSocketMessageHandler) subscribeAndSendMessageToCloud(conn *websocket.Conn, clientID string) {
+	nc := bridgemodel.GetNatsConnection()
+	subject := fmt.Sprintf("%s.>", msgs.NATSSYNC_MESSAGE_PREFIX)
+	_, err := nc.Subscribe(subject, func(msg *nats.Msg) {
+		log.Info("Received NATS message to send to cloud via websocket")
+		parsedSubject, err := msgs.ParseSubject(msg.Subject)
+		if err != nil {
+			log.WithError(err).Error("Failure to parse subject")
+			return
+		}
+		if parsedSubject.LocationID == clientID {
+			return
+		}
+
+		msgFormat := msgs.GetMsgFormat()
+		status, err := msgFormat.ValidateMsgFormat(msg.Data, false)
+		if err != nil {
+			log.Errorf("Error validating the cloud event message: %s", err.Error())
+			return
+		}
+		if !status {
+			log.Errorf("Cloud event message validation failed, ignoring the message...")
+			return
+		}
+
+		natmsg := bridgemodel.NatsMessage{Reply: msg.Reply, Subject: msg.Subject, Data: msg.Data}
+		envelope, enverr := msgs.PutObjectInEnvelope(natmsg, clientID, pkg.CLOUD_ID)
+		if enverr != nil {
+			log.Errorf("Error putting msg in envelope %s", enverr.Error())
+			return
+		}
+		jsonbits, jsonerr := json.Marshal(&envelope)
+		if jsonerr != nil {
+			log.Errorf("Error encoding envelope to json bits, wkipping message %s", jsonerr.Error())
+			return
+		}
+		var bmsgs []v1.BridgeMessage
+		bmsg := v1.BridgeMessage{ClientID: clientID, MessageData: string(jsonbits), FormatVersion: "1"}
+
+		request := v1.BridgeMessagePostReq{
+			AuthChallenge: *msgs.NewAuthChallengeFromStoredKey(),
+			Messages:      append(bmsgs, bmsg),
+		}
+
+		msgJSON, err := json.Marshal(request)
+		if err != nil {
+			log.WithError(err).Error("Failed to marshal msg to JSON")
+			return
+		}
+		if err = conn.WriteMessage(websocket.TextMessage, msgJSON); err != nil {
+			log.WithError(err).Error("Failed to send message to websocket")
+			return
+		}
+		log.Info("Message sent to cloud via websocket")
+	})
+	if err != nil {
+		log.WithError(err).Error("Failed to subscribe to subject")
+		return
+	}
+}
+
+func(t *WebSocketMessageHandler) ReadWSFromCloud(conn *websocket.Conn) {
+	nc := bridgemodel.GetNatsConnection()
```

Comment:
I see we have to do this call in most functions. It would probably be easier and more testable to pass in a NATS connection ptr when creating the handlers, and saving that ptr on the struct.